### ISSUE TITLE
Reading negative SByte values from Binary DataFormat returns 0

### ISSUE THREAD

**Issue by Merlin-san** [TeamSirenix/odin-serializer#40]
Hi, there is a minor bug with the BinaryDataReader which makes all negative `sbyte` values get read as 0. The repro is simple:

```cs
struct SerializedValue
{
    public sbyte value;
}

// And then in a method
byte[] bytes = SerializationUtility.SerializeValue<SerializedValue>(new SerializedValue() { value = -1 }, DataFormat.Binary);
Debug.Log(SerializationUtility.DeserializeValue<SerializedValue>(bytes, DataFormat.Binary).value);
```
This will log `0` in Unity instead of the expected value of `-1`. Switching to use JSON data format logs `-1` as expected. The above code only uses the struct since JSON doesn't serialize the plain sbyte without the container type.

The cause of this issue is that the BinaryDataWriter class will convert an `sbyte` to a `byte` in an unchecked block in the `WriteSByte` method and serialize that byte value directly, but the BinaryDataReader will read that value in `ReadSByte` from a `long` and cast it to an `sbyte` inside a `checked` block. This means that if for instance you store a `sbyte` value of `-1`, it will be serialized as the unsigned byte value of 255. Then when you attempt to read the value it will attempt to cast that `long` value of 255 to an `sbyte` value of `-1` which will throw an exception in the checked block since 255 would overflow to get to `-1` which gets caught and makes the method return 0.

An easy fix for this is would be to change the [`checked`](https://github.com/TeamSirenix/odin-serializer/blob/9885da859f2331ccf7bad617f1837756fbf09cc5/OdinSerializer/Core/DataReaderWriters/Binary/BinaryDataReader.cs#L691) block to an `unchecked` block in the `ReadSByte()` function in the `BinaryDataReader`. But that means any values which are supposed to be other types that fall out of the range of `sbyte` will just overflow into odd `sbyte` values. I didn't make a PR for this since I'm not sure how you'd want to approach this since `sbyte` is handled a bit differently than the other types.

**Comment by TheLastRar** [comment]
The issue seems to originate in ReadInt64

https://github.com/TeamSirenix/odin-serializer/blob/9885da859f2331ccf7bad617f1837756fbf09cc5/OdinSerializer/Core/DataReaderWriters/Binary/BinaryDataReader.cs#L920-L938

Here it is calling UNSAFE_Read_1_Byte for both bytes and sbytes
It should instead be calling UNSAFE_Read_1_SByte for the sbyte types as that function performs the necessary conversion to correctly read a negative sbyte

The separate cases for signed and unsigned data types is also what is done for other types (like short and int) so it seems like sbyte and byte should also have separate cases

**Comment by TorVestergaard** [comment]
Great catch, guys - this must have slipped in during the major unsafe code performance optimizations made to the binary format late last year. It should be fixed in commit 7aa69a03354e1aa4171ed2f6b0f0624c7d7b2463 now :)
